### ISSUE TITLE
weaver uses 100% cpu in one thread

### ISSUE THREAD

**Issue by lsorensen-extreme** [weaveworks/weave#3774]
## What you expected to happen?
weaver process to use minimal CPU when nothing is happening.

## What happened?
One thread is always using 100% cpu.
Using strace it shows that there is a call to select with an invalid timeout so it instantly returns and is called again and again such as:
[pid 15665] select(11, [10], NULL, NULL, {tv_sec=0, tv_usec=18446744072197722252}) = -1 EINVAL (Invalid argument)
Using delve it appears to be in a thread calling netpoll(block=true) which seems to result in the timeout being something quite invalid and hence does not block but instead instantly returns causing an infinite look of polling.

## How to reproduce it?
Install weave-net on a kubernetes cluster.  I was doing it on a ubuntu 18.04 system as well as on a custom build yocto and both did the same thing.  I tried both the current stable release as well as using 'latest' for the docker containers.  No difference.

## Anything else we need to know?
Fresh kubernetes install on a ubuntu host and then add weave-net to it usign kubectl create -f "https://cloud.weave.works/k8s/net?k8s-version=$(kubectl version | base64 | tr -d '\n')"

## Versions:
$ weave version
weave 2.6.0

$ docker version
Client:
 Version:           18.09.7
 API version:       1.39
 Go version:        go1.10.1
 Git commit:        2d0083d
 Built:             Fri Aug 16 14:20:06 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          18.09.7
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.10.1
  Git commit:       2d0083d
  Built:            Wed Aug 14 19:41:23 2019
  OS/Arch:          linux/amd64
  Experimental:     false

$ uname -a
Linux kube-master 4.15.0-76-generic #86-Ubuntu SMP Fri Jan 17 17:24:28 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux

$ kubectl version
Client Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.2", GitCommit:"c97fe5036ef3df2967d086711e6c0c405941e14b", GitTreeState:"clean", BuildDate:"2019-10-15T19:18:23Z", GoVersion:"go1.12.10", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.7", GitCommit:"be3d344ed06bff7a4fc60656200a93c74f31f9a4", GitTreeState:"clean", BuildDate:"2020-02-11T19:24:46Z", GoVersion:"go1.13.6", Compiler:"gc", Platform:"linux/amd64"}
```

## Logs:
$ kubectl logs -n kube-system <weave-net-pod> weave
Defaulting container name to weave.
Use 'kubectl describe pod/weave-net-t77xh -n kube-system' to see all of the containers in this pod.
OCI runtime exec failed: exec failed: container_linux.go:345: starting container process caused "exec: \"weave\": executable file not found in $PATH": unknown
command terminated with exit code 126
test@kube-master:~/weave$ kubectl logs -n kube-system weave-net-t77xh weave 
DEBU: 2020/02/14 20:50:00.049532 [kube-peers] Checking peer "56:f7:bc:83:1f:b1" against list &{[{9e:be:10:2b:f5:51 eca.extremenetworks.com} {56:f7:bc:83:1f:b1 kube-master} {a2:53:12:01:da:e3 eca2.extremenetworks.com}]}
INFO: 2020/02/14 20:50:00.099031 Command line options: map[conn-limit:200 datapath:datapath db-prefix:/weavedb/weave-net docker-api: expect-npc:true host-root:/host http-addr:127.0.0.1:6784 ipalloc-init:consensus=2 ipalloc-range:10.32.0.0/12 metrics-addr:0.0.0.0:6782 name:56:f7:bc:83:1f:b1 nickname:kube-master no-dns:true no-fastdp:true port:6783]
INFO: 2020/02/14 20:50:00.099073 weave  2.6.0
INFO: 2020/02/14 20:50:00.186068 Re-exposing 10.44.0.0/12 on bridge "weave"
INFO: 2020/02/14 20:50:00.212216 Re-exposing 10.32.0.1/12 on bridge "weave"
INFO: 2020/02/14 20:50:00.257588 Bridge type is bridge
INFO: 2020/02/14 20:50:00.257606 Communication between peers is unencrypted.
INFO: 2020/02/14 20:50:00.300334 Our name is 56:f7:bc:83:1f:b1(kube-master)
INFO: 2020/02/14 20:50:00.300380 Launch detected - using supplied peer list: [192.168.3.240 192.168.3.40]
INFO: 2020/02/14 20:50:00.309344 Checking for pre-existing addresses on weave bridge
INFO: 2020/02/14 20:50:00.309694 weave bridge has address 10.44.0.0/12
INFO: 2020/02/14 20:50:00.309718 weave bridge has address 10.32.0.1/12
INFO: 2020/02/14 20:50:00.328026 Found address 10.36.0.0/12 for ID _
INFO: 2020/02/14 20:50:00.328175 Found address 10.36.0.0/12 for ID _
INFO: 2020/02/14 20:50:00.329698 [allocator 56:f7:bc:83:1f:b1] Initialising with persisted data
INFO: 2020/02/14 20:50:00.329756 Sniffing traffic on vethwe-pcap (via pcap)
INFO: 2020/02/14 20:50:00.376487 ->[192.168.3.240:6783] attempting connection
INFO: 2020/02/14 20:50:00.376687 ->[192.168.3.40:6783] attempting connection
INFO: 2020/02/14 20:50:00.379104 ->[192.168.3.40:6783|a2:53:12:01:da:e3(eca2.extremenetworks.com)]: connection ready; using protocol version 2
INFO: 2020/02/14 20:50:00.388327 overlay_switch ->[a2:53:12:01:da:e3(eca2.extremenetworks.com)] using sleeve
INFO: 2020/02/14 20:50:00.388473 ->[192.168.3.40:6783|a2:53:12:01:da:e3(eca2.extremenetworks.com)]: connection added (new peer)
INFO: 2020/02/14 20:50:00.388616 ->[192.168.3.240:6783|9e:be:10:2b:f5:51(eca.extremenetworks.com)]: connection ready; using protocol version 2
INFO: 2020/02/14 20:50:00.388737 overlay_switch ->[9e:be:10:2b:f5:51(eca.extremenetworks.com)] using sleeve
INFO: 2020/02/14 20:50:00.388837 ->[192.168.3.240:6783|9e:be:10:2b:f5:51(eca.extremenetworks.com)]: connection added (new peer)
INFO: 2020/02/14 20:50:00.406745 Listening for HTTP control messages on 127.0.0.1:6784
INFO: 2020/02/14 20:50:00.406986 Listening for metrics requests on 0.0.0.0:6782
INFO: 2020/02/14 20:50:00.489034 ->[192.168.3.240:6783|9e:be:10:2b:f5:51(eca.extremenetworks.com)]: connection fully established
INFO: 2020/02/14 20:50:00.489115 ->[192.168.3.40:6783|a2:53:12:01:da:e3(eca2.extremenetworks.com)]: connection fully established
INFO: 2020/02/14 20:50:00.491903 sleeve ->[192.168.3.40:6783|a2:53:12:01:da:e3(eca2.extremenetworks.com)]: Effective MTU verified at 1438
INFO: 2020/02/14 20:50:00.491980 sleeve ->[192.168.3.240:6783|9e:be:10:2b:f5:51(eca.extremenetworks.com)]: Effective MTU verified at 1438
INFO: 2020/02/14 20:50:01.110052 [kube-peers] Added myself to peer list &{[{9e:be:10:2b:f5:51 eca.extremenetworks.com} {56:f7:bc:83:1f:b1 kube-master} {a2:53:12:01:da:e3 eca2.extremenetworks.com}]}

**Comment by lsorensen-extreme** [comment]
That value passed to select is not random.  Every system I had tried is passing the same invalid tv_usec value.  Not sure if the value has any meaning.  In hex it is FFFFFFFFA5E3508C which makes me think of a pointer or something close to -1 but not quite.

**Comment by bboreham** [comment]
Thanks for reporting this.
All calls to select, poll, etc., are done by the Go runtime, as far as I know.
Can you get a Go profile?  Should be on http://127.0.0.1:6784/debug/pprof/profile (curl and pipe to a file).

**Comment by lsorensen-extreme** [comment]
Here is the debug file.

[pprof.debug.profile.zip](https://github.com/weaveworks/weave/files/4225688/pprof.debug.profile.zip)

**Comment by uablrek** [comment]
I get the exact same problem (also checked with strace) but my environment is different;
```
Weave 2.6.1
K8s versions tested; v1.17.3, v1.18.0-beta.1 and local build v1.19.0-alpha.0.721+f8ff8f44206ff4
Linux kernel; 5.5.7
```
No alarming or repeated log printouts. Traffic works.

**Comment by bboreham** [comment]
@lsorensen-extreme sorry for delay.
Your profile shows the Weave Net program spending pretty much all of the time in pcap.
Hard to know what to say about that; it doesn't seem to be doing anything with the packets so perhaps they are not intended for Weave Net?
You can ask it to log more about what it sees with `--pkt-debug`  (but take care because it may be seeing a lot of packets hence generate a lot of logs).

![image](https://user-images.githubusercontent.com/8125524/76336563-17d82080-62ee-11ea-865d-1ffbbb81e09f.png)

**Comment by lsorensen-extreme** [comment]
I have almost no traffic.  Given strace says that select is being called with an invalid timeout, I suspect the reason it is spending time in pcap is because it is constantly trying to call WaitForPacket, but it returns immediately due to the invalid timeout specified.  Most likely the timeout value for the blocking case is calculated incorrectly resulting in garbage passed to the C code.  Not sure, since I am having a hard time following how the go runtime does things.

**Comment by uablrek** [comment]
Why is "pcap" used at all? It is a packet-tracing lib. Is there any debug/trace that can be turned off?

**Comment by bboreham** [comment]
@uablrek see https://github.com/weaveworks/weave/blob/master/docs/architecture.txt

Note that pcap is not used in the default "fast datapath" mode; only when Weave Net detects that the kernel doesn't support that mode does it fall back to pcap.

**Comment by bboreham** [comment]
Now I remember I've seen this before: https://github.com/google/gopacket/issues/708
So we need a change like this: https://github.com/weaveworks/scope/commit/9208e08bf3eaab9d276c53af6db5e9f7d141b4fb

**Comment by uablrek** [comment]
Good finding! Seems exactly to be the problem.

I can't see '-no-fastdp:true` in my log, but I can see this;
```
WARN: 2020/03/10 17:20:18.310825 Skipping bridge creation of "bridged_fastdp" due to: : bridge not supported
INFO: 2020/03/10 17:20:18.371930 Bridge type is bridge
```
What must be supported in the kernel for fastdp?

I build my kernel so I can easily enable what's needed. A problem I often have with K8s is that it expects a module and I often compile a function directly in the kernel. It may be that the function exist but not as a module. Not saying weave does that, just mentioning it :smile:

**Comment by bboreham** [comment]
That is a somewhat confusing error message, perhaps worth filing a bug about.
However it appears to come about if this happens: https://github.com/weaveworks/go-odp/blob/6b0aa22550d9325eb8f43418185859e13dc0de1d/odp/dpif.go#L58

It will attempt to load the `openvswitch` module but also try other things; you can see what is attempted in the surrounding code.

**Comment by lsorensen-extreme** [comment]
I purposely disabled fastdp due to some other network stuff that might interfere with it.  The timeout value integer conversion certainly seems likely.  I was suspecting it to be something like that happening.

**Comment by uablrek** [comment]
Thanks. I will make a try to add OVS. But if it takes too much time I wait for a fix for the pcap problem.

**Comment by bboreham** [comment]
Closed by #3782
